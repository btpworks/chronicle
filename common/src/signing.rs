use custom_error::custom_error;
use k256::{
    ecdsa::{SigningKey, VerifyingKey},
    pkcs8::FromPrivateKey,
    SecretKey,
};
use pkcs8::{FromPublicKey, ToPrivateKey};
use rand::prelude::StdRng;
use rand_core::SeedableRng;

use std::{
    path::{Path, PathBuf},
    string::FromUtf8Error,
};

use crate::models::AgentId;

custom_error! {pub SignerError
    Io{source: std::io::Error}                              = "Invalid key store directory",
    Pattern{source: glob::PatternError}                     = "Invalid glob ",
    Encoding{source: FromUtf8Error}                         = "Invalid file encoding",
    InvalidPublicKey{source: k256::pkcs8::Error}            = "Invalid public key",
    NoPublicKeyFound{}                                      = "No public key found",
    NoPrivateKeyFound{}                                     = "No private key found",
}

pub struct DirectoryStoredKeys {
    base: PathBuf,
}

impl DirectoryStoredKeys {
    pub fn new<P>(base: P) -> Result<Self, SignerError>
    where
        P: AsRef<Path>,
    {
        Ok(Self {
            base: base.as_ref().to_path_buf(),
        })
    }

    pub fn chronicle_signing(&self) -> Result<SigningKey, SignerError> {
        Self::signing_key_at(&self.base)
    }

    pub fn agent_signing(&self, agent: &AgentId) -> Result<SigningKey, SignerError> {
        Self::signing_key_at(&self.agent_path(agent))
    }

    /// If we have a signing key, derive the verifying key from it, else attempt to load an imported verifying key
    pub fn agent_verifying(&self, agent: &AgentId) -> Result<VerifyingKey, SignerError> {
        Self::signing_key_at(&self.agent_path(agent))
            .map(|signing| signing.verifying_key())
            .or_else(|_| Self::verifying_key_at(&self.agent_path(&agent)))
    }

    pub fn import_agent(
        &self,
        agent: &AgentId,
        signing: Option<&Path>,
        verifying: Option<&Path>,
    ) -> Result<(), SignerError> {
        if let Some(signing) = signing {
            std::fs::copy(
                Path::new(signing),
                Path::join(&self.agent_path(agent), Path::new("key.priv.pem")),
            )?;
        }

        if let Some(verifying) = verifying {
            std::fs::copy(
                Path::new(verifying),
                Path::join(&self.agent_path(agent), Path::new("key.pub.pem")),
            )?;
        }

        Ok(())
    }

    pub fn import_chronicle(
        &self,
        signing: Option<&Path>,
        verifying: Option<&Path>,
    ) -> Result<(), SignerError> {
        if let Some(signing) = signing {
            std::fs::copy(
                Path::new(signing),
                Path::join(&self.base, Path::new("key.priv.pem")),
            )?;
        }

        if let Some(verifying) = verifying {
            std::fs::copy(
                Path::new(verifying),
                Path::join(&self.base, Path::new("key.pub.pem")),
            )?;
        }

        Ok(())
    }

    pub fn generate_agent(&self, agent: &AgentId) -> Result<(), SignerError> {
        let path = self.agent_path(&agent);
        std::fs::create_dir_all(&path)?;
        Ok(Self::new_signing_key(&path)?)
    }

    pub fn generate_chronicle(&self) -> Result<(), SignerError> {
        std::fs::create_dir_all(&self.base)?;
        Ok(Self::new_signing_key(&self.base)?)
    }

    fn new_signing_key(secretpath: &Path) -> Result<(), SignerError> {
        let secret = SecretKey::random(StdRng::from_entropy());

        let privpem = secret.to_pkcs8_pem()?;

        std::fs::write(
            Path::join(Path::new(&secretpath), Path::new("key.priv.pem")),
            privpem.as_bytes(),
        )?;

        Ok(())
    }

    fn agent_path(&self, agent: &AgentId) -> PathBuf {
        Path::join(&self.base, Path::new(agent.decompose()))
    }

    fn signing_key_at(path: &Path) -> Result<SigningKey, SignerError> {
        Ok(SigningKey::read_pkcs8_pem_file(Path::join(
            path,
            Path::new("key.priv.pem"),
        ))?)
    }

    fn verifying_key_at(path: &Path) -> Result<VerifyingKey, SignerError> {
        Ok(VerifyingKey::read_public_key_pem_file(Path::join(
            path,
            Path::new("key.pub.pem"),
        ))?)
    }
}
