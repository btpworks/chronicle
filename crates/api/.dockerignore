sqlite_test
db.sqlite
