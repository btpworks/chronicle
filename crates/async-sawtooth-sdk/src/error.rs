use std::string::FromUtf8Error;

use futures::channel::mpsc::SendError;
use protobuf::ProtobufError;
use sawtooth_sdk::messaging::stream::ReceiveError;
use thiserror::Error;

use crate::ledger::BlockIdError;

#[derive(Error, Debug)]
pub enum SawtoothCommunicationError {
    #[error("ZMQ error: {0}")]
    ZMQ(#[from] zmq::Error),

    #[error("Send error: {0}")]
    SendChannel(#[from] SendError),

    #[error("Sawtooth receive error: {0}")]
    ReceiveSawtooth(#[from] ReceiveError),

    #[error("Sawtooth send error: {0}")]
    SendSawtooth(#[from] sawtooth_sdk::messaging::stream::SendError),

    #[error("Protobuf error: {0}")]
    Protobuf(#[from] ProtobufError),
    #[error("Protobuf decode error: {0}")]
    ProtobufProst(#[from] prost::DecodeError),
    #[error("Unexpected Status: {status:?}")]
    UnexpectedStatus { status: i32 },
    #[error("No transaction id for event")]
    MissingTransactionId,
    #[error("Cannot determine block number for event")]
    MissingBlockNum,
    #[error("Cannot determine block id for event")]
    MissingBlockId,
    #[error("Unexpected message structure")]
    MalformedMessage,
    #[error("Json: {0}")]
    Json(#[from] serde_json::Error),
    #[error("Subscribe error: {code}")]
    SubscribeError { code: i32 },
    #[error("Block number is not number: {source}")]
    BlockNumNotNumber {
        #[from]
        source: std::num::ParseIntError,
    },
    #[error("No blocks returned when searching for current block")]
    NoBlocksReturned,
    #[error("Ledger event parse error: {source}")]
    LedgerEventParse {
        #[from]
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[error("Utf8 {source}")]
    Utf8 {
        #[from]
        source: FromUtf8Error,
    },

    #[error("BlockId {source}")]
    BlockId {
        #[from]
        source: BlockIdError,
    },

    #[error("Infallible {0}")]
    Infallible(#[from] std::convert::Infallible),
}
