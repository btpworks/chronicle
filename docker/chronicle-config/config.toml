[secrets]
path = "/var/lib/chronicle/secrets"
[validator]
address = "tcp://validator:4004"
[namespace_bindings]
