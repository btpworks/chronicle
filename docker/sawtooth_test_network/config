          [secrets]
   3   │  path = "~/.chronicle/secrets"
   4   │  [store]
   5   │  path = "~/.chronicle/store"
   6   │  [validator]
   7   │  address = "tcp://validator:4004"