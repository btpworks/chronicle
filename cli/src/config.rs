use super::CliError;
use clap::ArgMatches;
use common::signing::DirectoryStoredKeys;


use question::{Answer, Question};


use std::path::{Path, PathBuf};
use url::Url;

#[derive(Serialize, Deserialize, Debug)]
pub struct SecretConfig {
    pub path: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct StoreConfig {
    pub path: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ValidatorConfig {
    pub address: Url,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Config {
    pub secrets: SecretConfig,
    pub store: StoreConfig,
    pub validator: ValidatorConfig,
}

pub fn handle_config_and_init(matches: &ArgMatches) -> Result<Config, CliError> {
    let path = matches.value_of("config").unwrap().to_owned();
    let path = shellexpand::tilde(&path);
    let path = PathBuf::from(&*path);

    let meta = std::fs::metadata(&*path);
    if meta.is_err() {
        init_chronicle_at(&path)?
    }

    let toml = std::fs::read_to_string(&path)?;
    Ok(toml::from_str(&toml)?)
}

/// Interrogate the user for required configuration and initialise chronicle to a working state, including key generation if needed
fn init_chronicle_at(path: &Path) -> Result<(), CliError> {
    let init = Question::new(&format!(
        "No configuration found at {}, create?",
        path.to_string_lossy()
    ))
    .default(question::Answer::YES)
    .show_defaults()
    .confirm();

    if init != Answer::YES {
        std::process::exit(0);
    }

    let dbpath = Question::new("Where should chronicle store state?")
        .default(Answer::RESPONSE(
            Path::join(
                path.parent().ok_or(CliError::InvalidPath)?,
                PathBuf::from("store"),
            )
            .to_string_lossy()
            .to_string(),
        ))
        .show_defaults()
        .confirm();

    let secretpath = Question::new("Where should chronicle store secrets?")
        .default(Answer::RESPONSE(
            Path::join(
                path.parent().ok_or(CliError::InvalidPath)?,
                PathBuf::from("secrets"),
            )
            .to_string_lossy()
            .to_string(),
        ))
        .show_defaults()
        .confirm();

    let validatorurl =
        Question::new("What is the address of the sawtooth validator zeromq service?")
            .default(Answer::RESPONSE("tcp://localhost:4004".to_owned()))
            .show_defaults()
            .confirm();

    let generatesecret = Question::new("Generate a new default key in the secret store?")
        .default(Answer::YES)
        .show_defaults()
        .confirm();

    match (dbpath, secretpath, validatorurl) {
        (
            Answer::RESPONSE(dbpath),
            Answer::RESPONSE(secretpath),
            Answer::RESPONSE(validatorurl),
        ) => {
            let dbpath = Path::new(&dbpath);
            let secretpath = Path::new(&secretpath);

            println!("Creating config dir {} if needed", path.to_string_lossy());
            std::fs::create_dir_all(path.parent().unwrap())?;
            println!("Creating db dir {} if needed", &dbpath.to_string_lossy());
            std::fs::create_dir_all(&dbpath)?;
            println!(
                "Creating secret dir {} if needed",
                &secretpath.to_string_lossy()
            );
            std::fs::create_dir_all(&secretpath)?;

            let config = format!(
                r#"
            [secrets]
            path = "{}"
            [store]
            path = "{}"
            [validator]
            address = "{}"
            "#,
                &*secretpath.to_string_lossy(),
                &*dbpath.to_string_lossy(),
                validatorurl
            );

            println!("Writing config to {}", &path.to_string_lossy());
            println!("{}", &config);

            std::fs::write(path, config)?;

            if generatesecret == Answer::YES {
                DirectoryStoredKeys::new(&secretpath)?.generate_chronicle()?;
            } else {
                println!(
                    "Please install your keys in .pem format in the configured secret location"
                );
                std::process::exit(0);
            }
        }
        _ => unreachable!(),
    }

    Ok(())
}
