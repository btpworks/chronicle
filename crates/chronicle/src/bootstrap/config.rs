use super::{CliError, CliModel, SubCommand};
use common::signing::DirectoryStoredKeys;
use question::{Answer, Question};
use serde_derive::{Deserialize, Serialize};
use uuid::Uuid;

use std::{
    collections::HashMap,
    path::{Path, PathBuf},
};
use url::Url;

#[derive(Serialize, Deserialize, Debug)]
pub(crate) struct SecretConfig {
    pub path: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub(crate) struct ValidatorConfig {
    pub address: Url,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Config {
    pub(crate) secrets: SecretConfig,
    pub(crate) validator: ValidatorConfig,
    pub(crate) namespace_bindings: HashMap<String, Uuid>,
}

pub(crate) fn handle_config_and_init(model: &CliModel) -> Result<Config, CliError> {
    let path = model
        .as_cmd()
        .get_matches()
        .get_one::<String>("config")
        .unwrap()
        .to_owned();
    let path = shellexpand::tilde(&path);
    let path = PathBuf::from(&*path);

    let meta = std::fs::metadata(&*path);
    if meta.is_err() {
        init_chronicle_at(&path)?
    }

    let toml = std::fs::read_to_string(&path)?;
    Ok(toml::from_str(&toml)?)
}

/// Interrogate the user for required configuration and initialise chronicle to a working state, including key generation if needed
fn init_chronicle_at(path: &Path) -> Result<(), CliError> {
    let init = Question::new(&format!(
        "No configuration found at {}, create?",
        path.to_string_lossy()
    ))
    .default(question::Answer::YES)
    .show_defaults()
    .confirm();

    if init != Answer::YES {
        std::process::exit(0);
    }

    let secretpath = Question::new("Where should Chronicle store secrets?")
        .default(Answer::RESPONSE(
            Path::join(
                path.parent().ok_or(CliError::InvalidPath {
                    path: path.to_string_lossy().to_string(),
                })?,
                PathBuf::from("secrets"),
            )
            .to_string_lossy()
            .to_string(),
        ))
        .show_defaults()
        .confirm();

    let validatorurl =
        Question::new("What is the address of the sawtooth validator zeromq service?")
            .default(Answer::RESPONSE("tcp://localhost:4004".to_owned()))
            .show_defaults()
            .confirm();

    let generatesecret = Question::new("Generate a new default key in the secret store?")
        .default(Answer::YES)
        .show_defaults()
        .confirm();

    match (secretpath, validatorurl) {
        (Answer::RESPONSE(secretpath), Answer::RESPONSE(validatorurl)) => {
            let secretpath = Path::new(&secretpath);

            println!("Creating config dir {} if needed", path.to_string_lossy());
            std::fs::create_dir_all(path.parent().unwrap())?;
            println!(
                "Creating secret dir {} if needed",
                &secretpath.to_string_lossy()
            );
            std::fs::create_dir_all(secretpath)?;

            let config = format!(
                r#"[secrets]
path = "{}"
[validator]
address = "{}"
[namespace_bindings]
"#,
                &*secretpath.to_string_lossy(),
                validatorurl
            );

            println!("Writing config to {}", &path.to_string_lossy());
            println!("{}", &config);

            std::fs::write(path, config)?;

            if generatesecret == Answer::YES {
                DirectoryStoredKeys::new(secretpath)?.generate_chronicle()?;
            } else {
                println!(
                    "Please install your keys in .pem format in the configured secret location"
                );
                std::process::exit(0);
            }
        }
        _ => unreachable!(),
    }

    Ok(())
}
