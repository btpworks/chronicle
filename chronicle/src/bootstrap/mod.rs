mod cli;
mod config;
pub mod telemetry;

use api::{
    chronicle_graphql::{ChronicleGraphQl, ChronicleGraphQlServer},
    Api, ApiDispatch, ApiError, ConnectionOptions, UuidGen,
};
use async_graphql::ObjectType;
use clap::{ArgMatches, Command};
use clap_complete::{generate, Generator, Shell};
pub use cli::*;
use common::{
    attributes::Attributes,
    commands::{
        ActivityCommand, AgentCommand, ApiCommand, ApiResponse, EntityCommand, KeyImport,
        KeyRegistration, NamespaceCommand, PathOrFile, QueryCommand,
    },
    prov::{ActivityId, AgentId, CompactionError, DomaintypeId, EntityId},
};
use custom_error::custom_error;
use tokio::sync::broadcast::error::RecvError;
use tracing::{error, instrument};
use user_error::UFE;

use common::signing::SignerError;
use config::*;
use diesel::{
    r2d2::{ConnectionManager, Pool},
    SqliteConnection,
};

use sawtooth_protocol::{events::StateDelta, messaging::SawtoothSubmitter};
use url::Url;

use std::{
    io,
    net::SocketAddr,
    path::{Path, PathBuf},
    time::Duration,
};

#[allow(dead_code)]
fn submitter(config: &Config, options: &ArgMatches) -> Result<SawtoothSubmitter, SignerError> {
    Ok(SawtoothSubmitter::new(
        &options
            .value_of("sawtooth")
            .map(Url::parse)
            .unwrap_or_else(|| Ok(config.validator.address.clone()))?,
        &common::signing::DirectoryStoredKeys::new(&config.secrets.path)?.chronicle_signing()?,
    ))
}

#[allow(dead_code)]
fn state_delta(config: &Config, options: &ArgMatches) -> Result<StateDelta, SignerError> {
    Ok(StateDelta::new(
        &options
            .value_of("sawtooth")
            .map(Url::parse)
            .unwrap_or_else(|| Ok(config.validator.address.clone()))?,
        &common::signing::DirectoryStoredKeys::new(&config.secrets.path)?.chronicle_signing()?,
    ))
}

#[cfg(feature = "inmem")]
fn ledger() -> Result<common::ledger::InMemLedger, std::convert::Infallible> {
    Ok(common::ledger::InMemLedger::new())
}

#[derive(Debug, Clone)]
struct UniqueUuid;

impl UuidGen for UniqueUuid {}

type ConnectionPool = Pool<ConnectionManager<SqliteConnection>>;
fn pool(config: &Config) -> Result<ConnectionPool, ApiError> {
    Ok(Pool::builder()
        .connection_customizer(Box::new(ConnectionOptions {
            enable_wal: true,
            enable_foreign_keys: true,
            busy_timeout: Some(Duration::from_secs(2)),
        }))
        .build(ConnectionManager::<SqliteConnection>::new(
            &*Path::join(&config.store.path, &PathBuf::from("db.sqlite")).to_string_lossy(),
        ))?)
}

fn graphql_addr(options: &ArgMatches) -> Result<Option<SocketAddr>, ApiError> {
    if !options.is_present("gql") {
        Ok(None)
    } else if let Some(addr) = options.value_of("gql-interface") {
        Ok(Some(addr.parse()?))
    } else {
        Ok(None)
    }
}

pub async fn graphql_server<Query, Mutation>(
    api: &ApiDispatch,
    pool: &ConnectionPool,
    gql: ChronicleGraphQl<Query, Mutation>,
    options: &ArgMatches,
    open: bool,
) -> Result<(), ApiError>
where
    Query: ObjectType + Copy,
    Mutation: ObjectType + Copy,
{
    if let Some(addr) = graphql_addr(options)? {
        gql.serve_graphql(pool.clone(), api.clone(), addr, open)
            .await
    }

    Ok(())
}

#[cfg(not(feature = "inmem"))]
pub async fn api(
    pool: &ConnectionPool,
    options: &ArgMatches,
    config: &Config,
) -> Result<ApiDispatch, ApiError> {
    let submitter = submitter(config, options)?;
    let state = state_delta(config, options)?;

    Api::new(
        pool.clone(),
        submitter,
        state,
        &config.secrets.path,
        UniqueUuid,
    )
    .await
}

#[cfg(feature = "inmem")]
pub async fn api(
    pool: &ConnectionPool,
    _options: &ArgMatches,
    config: &Config,
) -> Result<api::ApiDispatch, ApiError> {
    let mut ledger = ledger()?;
    let state = ledger.reader();

    Api::new(
        pool.clone(),
        ledger,
        state,
        &config.secrets.path,
        UniqueUuid,
    )
    .await
}

fn domain_type(args: &ArgMatches) -> Option<DomaintypeId> {
    if !args.is_present("domaintype") {
        None
    } else {
        args.value_of("domaintype").map(DomaintypeId::from_name)
    }
}

#[instrument(skip(gql))]
async fn execute_arguments<Query, Mutation>(
    gql: ChronicleGraphQl<Query, Mutation>,
    config: Config,
    options: &ArgMatches,
) -> Result<(ApiResponse, ApiDispatch), ApiError>
where
    Query: ObjectType + Copy,
    Mutation: ObjectType + Copy,
{
    dotenv::dotenv().ok();

    let pool = pool(&config)?;
    let api = api(&pool, options, &config).await?;
    let ret_api = api.clone();

    let execution = vec![
        options.subcommand_matches("namespace").and_then(|m| {
            m.subcommand_matches("create").map(|m| {
                api.dispatch(ApiCommand::NameSpace(NamespaceCommand::Create {
                    name: m.value_of("namespace").unwrap().into(),
                }))
            })
        }),
        options.subcommand_matches("agent").and_then(|m| {
            vec![
                m.subcommand_matches("create").map(|m| {
                    api.dispatch(ApiCommand::Agent(AgentCommand::Create {
                        name: m.value_of("agent_name").unwrap().into(),
                        namespace: m.value_of("namespace").unwrap().into(),
                        attributes: Attributes::type_only(domain_type(m)),
                    }))
                }),
                m.subcommand_matches("register-key").map(|m| {
                    let registration = {
                        if m.is_present("generate") {
                            KeyRegistration::Generate
                        } else if m.is_present("privatekey") {
                            KeyRegistration::ImportSigning(KeyImport::FromPath {
                                path: m.value_of_t::<PathBuf>("privatekey").unwrap(),
                            })
                        } else {
                            KeyRegistration::ImportVerifying(KeyImport::FromPath {
                                path: m.value_of_t::<PathBuf>("privatekey").unwrap(),
                            })
                        }
                    };

                    api.dispatch(ApiCommand::Agent(AgentCommand::RegisterKey {
                        id: AgentId::from_name(m.value_of("agent_name").unwrap()),
                        namespace: m.value_of("namespace").unwrap().into(),
                        registration,
                    }))
                }),
                m.subcommand_matches("use").map(|m| {
                    api.dispatch(ApiCommand::Agent(AgentCommand::UseInContext {
                        id: AgentId::from_name(m.value_of("agent_name").unwrap()),
                        namespace: m.value_of("namespace").unwrap().into(),
                    }))
                }),
            ]
            .into_iter()
            .flatten()
            .next()
        }),
        options.subcommand_matches("activity").and_then(|m| {
            vec![
                m.subcommand_matches("create").map(|m| {
                    api.dispatch(ApiCommand::Activity(ActivityCommand::Create {
                        name: m.value_of("activity_name").unwrap().into(),
                        namespace: m.value_of("namespace").unwrap().into(),
                        attributes: Attributes::type_only(domain_type(m)),
                    }))
                }),
                m.subcommand_matches("start").map(|m| {
                    api.dispatch(ApiCommand::Activity(ActivityCommand::Start {
                        id: ActivityId::from_name(m.value_of("activity_name").unwrap()),
                        namespace: m.value_of("namespace").unwrap().into(),
                        time: None,
                        agent: None,
                    }))
                }),
                m.subcommand_matches("end").map(|m| {
                    api.dispatch(ApiCommand::Activity(ActivityCommand::End {
                        id: m.value_of("activity_name").map(ActivityId::from_name),
                        namespace: m.value_of("namespace").unwrap().into(),
                        time: None,
                        agent: None,
                    }))
                }),
                m.subcommand_matches("use").map(|m| {
                    api.dispatch(ApiCommand::Activity(ActivityCommand::Use {
                        id: EntityId::from_name(m.value_of("entity_name").unwrap()),
                        namespace: m.value_of("namespace").unwrap().into(),
                        activity: m.value_of("activity_name").map(ActivityId::from_name),
                    }))
                }),
                m.subcommand_matches("generate").map(|m| {
                    api.dispatch(ApiCommand::Activity(ActivityCommand::Generate {
                        id: EntityId::from_name(m.value_of("entity_name").unwrap()),
                        namespace: m.value_of("namespace").unwrap().into(),
                        activity: m.value_of("activity_name").map(ActivityId::from_name),
                    }))
                }),
            ]
            .into_iter()
            .flatten()
            .next()
        }),
        options.subcommand_matches("entity").and_then(|m| {
            vec![m.subcommand_matches("attach").map(|m| {
                api.dispatch(ApiCommand::Entity(EntityCommand::Attach {
                    id: EntityId::from_name(m.value_of("entity_name").unwrap()),
                    namespace: m.value_of("namespace").unwrap().into(),
                    file: PathOrFile::Path(m.value_of_t::<PathBuf>("file").unwrap()),
                    locator: m.value_of("locator").map(|x| x.to_owned()),
                    agent: m.value_of("agent").map(AgentId::from_name),
                }))
            })]
            .into_iter()
            .flatten()
            .next()
        }),
        options.subcommand_matches("export").map(|m| {
            api.dispatch(ApiCommand::Query(QueryCommand {
                namespace: m.value_of("namespace").unwrap().to_owned(),
            }))
        }),
    ]
    .into_iter()
    .flatten()
    .next();

    // If we actually execute a command, then do not run the api
    if let Some(execution) = execution {
        let exresult = execution.await;

        Ok((exresult?, ret_api))
    } else {
        graphql_server(&api, &pool, gql, options, options.is_present("open")).await?;

        Ok((ApiResponse::Unit, ret_api))
    }
}

custom_error! {pub CliError
    Api{source: api::ApiError}                  = "Api error",
    Keys{source: SignerError}                   = "Key storage",
    FileSystem{source: std::io::Error}          = "Cannot locate configuration file",
    ConfigInvalid{source: toml::de::Error}      = "Invalid configuration file",
    InvalidPath                                 = "Invalid path", //TODO - the path, you know how annoying this is
    Ld{source: CompactionError}                 = "Invalid Json LD",
    CommitNoticiationStream {source: RecvError} = "Failure in commit notification stream"
}

impl UFE for CliError {}

async fn config_and_exec<Query, Mutation>(
    gql: ChronicleGraphQl<Query, Mutation>,
    matches: &ArgMatches,
) -> Result<(), CliError>
where
    Query: ObjectType + Copy,
    Mutation: ObjectType + Copy,
{
    use colored_json::prelude::*;
    let config = handle_config_and_init(matches)?;
    let response = execute_arguments(gql, config, matches).await?;

    match response {
        (
            ApiResponse::Submission {
                subject,
                prov: _,
                correlation_id,
            },
            api,
        ) => {
            // For commands that have initiated a ledger operation, wait for the matching result
            let mut tx_notifications = api.notify_commit.subscribe();

            loop {
                let (_prov, incoming_correlation_id) =
                    tx_notifications.recv().await.map_err(CliError::from)?;
                if correlation_id == incoming_correlation_id {
                    println!("{}", subject);
                    break;
                }
            }
        }
        (ApiResponse::QueryReply { prov }, _) => {
            println!(
                "{}",
                prov.to_json()
                    .compact()
                    .await?
                    .to_string()
                    .to_colored_json_auto()
                    .unwrap()
            );
        }
        (ApiResponse::Unit, _api) => {}
    };
    Ok(())
}

fn print_completions<G: Generator>(gen: G, app: &mut Command) {
    generate(gen, app, app.get_name().to_string(), &mut io::stdout());
}

pub async fn bootstrap<Query, Mutation>(gql: ChronicleGraphQl<Query, Mutation>)
where
    Query: ObjectType + 'static + Copy,
    Mutation: ObjectType + 'static + Copy,
{
    let matches = cli().get_matches();

    if let Ok(generator) = matches.value_of_t::<Shell>("completions") {
        let mut app = cli();
        eprintln!("Generating completion file for {}...", generator);
        print_completions(generator, &mut app);
        std::process::exit(0);
    }

    if matches.is_present("export-schema") {
        print!("{}", gql.exportable_schema());
        std::process::exit(0);
    }

    if matches.is_present("console-logging") {
        telemetry::console_logging();
    }

    if matches.is_present("instrument") {
        telemetry::telemetry(
            Url::parse(&*matches.value_of_t::<String>("instrument").unwrap()).unwrap(),
        );
    }

    config_and_exec(gql, &matches)
        .await
        .map_err(|e| {
            error!(?e, "Api error");
            e.into_ufe().print();
            std::process::exit(1);
        })
        .ok();

    std::process::exit(0);
}
