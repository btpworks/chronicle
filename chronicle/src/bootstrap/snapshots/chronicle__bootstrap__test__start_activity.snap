---
source: chronicle/src/bootstrap/mod.rs
expression: "serde_json::to_string_pretty(&v).unwrap()"
---
{
  "@context": {
    "@version": 1.1,
    "actedOnBehalfOf": {
      "@container": "@set",
      "@id": "prov:actedOnBehalfOf",
      "@type": "@id"
    },
    "activity": {
      "@id": "prov:activity",
      "@type": "@id"
    },
    "agent": {
      "@id": "prov:agent",
      "@type": "@id"
    },
    "attachment": {
      "@id": "chronicle:hasAttachment",
      "@type": "@id"
    },
    "chronicle": "http://blockchaintp.com/chronicle/ns#",
    "endTime": {
      "@id": "prov:endedAtTime"
    },
    "entity": {
      "@id": "prov:entity",
      "@type": "@id"
    },
    "hadPrimarySource": {
      "@container": "@set",
      "@id": "prov:hadPrimarySource",
      "@type": "@id"
    },
    "identity": {
      "@id": "chronicle:hasIdentity",
      "@type": "@id"
    },
    "label": {
      "@id": "rdfs:label"
    },
    "namespace": {
      "@id": "chronicle:hasNamespace",
      "@type": "@id"
    },
    "previousAttachments": {
      "@container": "@set",
      "@id": "chronicle:hadAttachment",
      "@type": "@id"
    },
    "previousIdentities": {
      "@container": "@set",
      "@id": "chronicle:hadIdentity",
      "@type": "@id"
    },
    "prov": "http://www.w3.org/ns/prov#",
    "provext": "https://openprovenance.org/ns/provext#",
    "publicKey": {
      "@id": "chronicle:hasPublicKey"
    },
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
    "signature": {
      "@id": "chronicle:entitySignature"
    },
    "signedAtTime": {
      "@id": "chronicle:signedAtTime"
    },
    "source": {
      "@id": "chronicle:entityLocator"
    },
    "startTime": {
      "@id": "prov:startedAtTime"
    },
    "used": {
      "@container": "@set",
      "@id": "prov:used",
      "@type": "@id"
    },
    "value": {
      "@id": "chronicle:value",
      "@type": "@json"
    },
    "wasAssociatedWith": {
      "@container": "@set",
      "@id": "prov:wasAssociatedWith",
      "@type": "@id"
    },
    "wasDerivedFrom": {
      "@container": "@set",
      "@id": "prov:wasDerivedFrom",
      "@type": "@id"
    },
    "wasGeneratedBy": {
      "@container": "@set",
      "@id": "prov:wasGeneratedBy",
      "@type": "@id"
    },
    "wasQuotedFrom": {
      "@container": "@set",
      "@id": "prov:wasQuotedFrom",
      "@type": "@id"
    },
    "wasRevisionOf": {
      "@container": "@set",
      "@id": "prov:wasRevisionOf",
      "@type": "@id"
    },
    "xsd": "http://www.w3.org/2001/XMLSchema#"
  },
  "@graph": [
    {
      "@id": "chronicle:activity:testactivity",
      "@type": "prov:Activity",
      "label": "testactivity",
      "namespace": "chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea",
      "startTime": "2014-07-08T09:10:11+00:00",
      "value": {},
      "wasAssociatedWith": [
        "chronicle:agent:testagent"
      ]
    },
    {
      "@id": "chronicle:agent:testagent",
      "@type": "prov:Agent",
      "label": "testagent",
      "namespace": "chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea",
      "value": {}
    },
    {
      "@id": "chronicle:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea",
      "@type": "chronicle:Namespace",
      "label": "testns"
    }
  ]
}
